kernel/src/history_manager/timestamp_visitor.rs

```diff
@@ -0,0 +1,137 @@
+//! This module defines a visitor used to extract the in-commit timestamp from a commit file if it
+//! is present.
+use std::sync::{Arc, LazyLock};
+
+use crate::actions::COMMIT_INFO_NAME;
+use crate::expressions::column_name;
+use crate::schema::{ColumnNamesAndTypes, DataType, Schema, StructField, StructType};
+use crate::utils::require;
+use crate::{DeltaResult, Error, RowVisitor};
+
+/// This visitor extracts the in-commit timestamp (ICT) from a CommitInfo action if present. The
+/// [`EngineData`] being visited must have the schema defined in
+/// [`InCommitTimestampVisitor::schema`].
+///
+/// Only one row of the engine data is checked because CommitInfo. This is because in-commit
+/// timestamps requires that the CommitInfo containing the ICT be the first action in the log.
+#[allow(unused)]
+#[derive(Default)]
+pub(crate) struct InCommitTimestampVisitor {
+    pub(crate) in_commit_timestamp: Option<i64>,
+}
+
+impl InCommitTimestampVisitor {
+    #[allow(unused)]
+    /// Get the schema that the visitor expects the data to have.
+    pub(crate) fn schema() -> Arc<Schema> {
+        static SCHEMA: LazyLock<Arc<Schema>> = LazyLock::new(|| {
+            let ict_type = StructField::new("inCommitTimestamp", DataType::LONG, true);
+            Arc::new(StructType::new(vec![StructField::new(
+                COMMIT_INFO_NAME,
+                StructType::new([ict_type]),
+                true,
+            )]))
+        });
+        SCHEMA.clone()
+    }
+}
+impl RowVisitor for InCommitTimestampVisitor {
+    fn selected_column_names_and_types(
+        &self,
+    ) -> (&'static [crate::schema::ColumnName], &'static [DataType]) {
+        static NAMES_AND_TYPES: LazyLock<ColumnNamesAndTypes> = LazyLock::new(|| {
+            let names = vec![column_name!("commitInfo.inCommitTimestamp")];
+            let types = vec![DataType::LONG];
+
+            (names, types).into()
+        });
+        NAMES_AND_TYPES.as_ref()
+    }
+
+    fn visit<'a>(
+        &mut self,
+        row_count: usize,
+        getters: &[&'a dyn crate::engine_data::GetData<'a>],
+    ) -> DeltaResult<()> {
+        require!(
+            getters.len() == 1,
+            Error::InternalError(format!(
+                "Wrong number of InCommitTimestampVisitor getters: {}",
+                getters.len()
+            ))
+        );
+
+        // If the batch is empty, return
+        if row_count == 0 {
+            return Ok(());
+        }
+        // CommitInfo must be the first action in a commit
+        if let Some(in_commit_timestamp) = getters[0].get_long(0, "commitInfo.inCommitTimestamp")? {
+            self.in_commit_timestamp = Some(in_commit_timestamp);
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use arrow_55::array::StringArray;
```

Comment:
interesting - didn't realize these symbols were getting exported (though I guess it makes sense)

instead we should use `crate::arrow::array::StringArray`